### PR TITLE
Bugfix/4449 dict attribute error

### LINKED ISSUE #3994: AttributeError when using multiple dataloaders
## 🐛 Bug

When training with multiple datasets and using `self.log()` for logging, I get following `AttributeError`:

`AttributeError: 'dict' object has no attribute 'get_epoch_log_metrics'`

### Proposed fix

After some code digging, I suppose [this line](https://github.com/PyTorchLightning/pytorch-lightning/blob/53e056612c1d27deb45916351222522efabfaf35/pytorch_lightning/trainer/connectors/logger_connector.py#L208) is the culprit. I'm guessing it should be replaced with:

```python
result = Result()
result.update({
    f'{k}/dataloader_idx_{dataloader_idx}': v
    for k, v in metrics.items()
})
```

or something similar.

### LINKED ISSUE #4449: AttributeError: 'dict' object has no attribute 'get_epoch_log_metrics'
####  I have created a lightning module which is working fine for single validation dataset but throws following error while using multiple validation dataset: 

```
self._log_on_evaluation_epoch_end_metrics(epoch_logs)
  File "/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/pytorch_lightning/trainer/connectors/logger_connector.py", line 186, in _log_on_evaluation_epoch_end_metrics
    logger_metrics = reduced_epoch_metrics.get_epoch_log_metrics()
AttributeError: 'dict' object has no attribute 'get_epoch_log_metrics'
```

Here's my `validation_step` method. To support single as well as multiple validation dataset I have kept `dataset_idx = 0`

```
    def validation_step(self, batch, batch_idx, dataset_idx=0):
        qa_pairs, labels = batch
        # labels = labels.type(torch.LongTensor)
        labels = labels.long()#(torch.LongTensor)
        # run the model for the inputs 
        outputs = self(qa_pairs)
        # compute the loss
        if self.loss.lower() == "cross_entropy":
            loss = F.cross_entropy(outputs, labels)

        self.valid_acc(outputs, labels)
        self.log('val_loss', loss, on_step=True)
        self.log('val_acc', self.valid_acc, on_step=True, on_epoch=True, prog_bar=True)
```

Looking at the error msgs, I commented the logging part, which seems to get rid of the error.  

#### system:
pytorch: 1.6
ubuntu 18.04 
pytorch-lightning: 1.0.3

[StackOverflow Post](https://stackoverflow.com/questions/64614497/attributeerror-dict-object-has-no-attribute-get-epoch-log-metrics-pytorch-l)

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolve 2 bugs related to logging, but one is still there. Reduction is not properly done on legacy results.

Fixes #4449
Fixes #3994

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -351,23 +389,11 @@ def __process_eval_epoch_end_results_and_log_legacy(self, eval_results, test_mod
...
-                # log metrics
-                if len(log_metrics) > 0:
-                    self.trainer.logger_connector.log_metrics(log_metrics, {})
-
-                # track metrics for callbacks (all prog bar, logged and callback metrics)
-                self.trainer.logger_connector.callback_metrics.update(callback_metrics)
-                self.trainer.logger_connector.callback_metrics.update(log_metrics)
-                self.trainer.logger_connector.callback_metrics.update(prog_bar_metrics)
+                if num_loaders > 1:
+                    self.__process_eval_epoch_end_results_and_log_legacy_udpate(prog_bar_metrics, log_metrics, callback_metrics)
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:393]:
```suggestion
                    self.__process_eval_epoch_end_results_and_log_legacy_update(prog_bar_metrics, log_metrics, callback_metrics)
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -329,6 +346,25 @@ def _track_callback_metrics(self, eval_results, using_eval_result):
                     flat['early_stop_on'] = flat['val_loss']
                 self.trainer.logger_connector.callback_metrics.update(flat)
 
+    def __process_eval_epoch_end_results_and_log_legacy_udpate(self, prog_bar_metrics, log_metrics, callback_metrics):
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:349]:
```suggestion
    def __process_eval_epoch_end_results_and_log_legacy_update(self, prog_bar_metrics, log_metrics, callback_metrics):
```

**Review by SeanNaren — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 3]
```diff
@@ -351,23 +389,11 @@ def __process_eval_epoch_end_results_and_log_legacy(self, eval_results, test_mod
...
-                self.trainer.logger_connector.callback_metrics.update(callback_metrics)
-                self.trainer.logger_connector.callback_metrics.update(log_metrics)
-                self.trainer.logger_connector.callback_metrics.update(prog_bar_metrics)
+                if num_loaders > 1:
+                    self.__process_eval_epoch_end_results_and_log_legacy_udpate(prog_bar_metrics, log_metrics, callback_metrics)
 
-                if len(dataloader_result_metrics) > 0:
-                    self.eval_loop_results.append(dataloader_result_metrics)
+            if num_loaders == 1:
+                self.__process_eval_epoch_end_results_and_log_legacy_udpate(prog_bar_metrics, log_metrics, callback_metrics)
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:396]: PyRight is throwing an error here as these variables can be unbound

```  
396:77 - error: "prog_bar_metrics" is possibly unbound (reportUnboundVariable)
396:95 - error: "log_metrics" is possibly unbound (reportUnboundVariable)
396:108 - error: "callback_metrics" is possibly unbound (reportUnboundVariable)
```

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]: Done !

**Review by SeanNaren — APPROVED** [review]

**Pull request merged**